core/contract/src/main/java/org/eclipse/dataspaceconnector/contract/negotiation/ProviderContractNegotiationManagerImpl.java

```diff
@@ -399,6 +414,7 @@ private int checkConfirming() {
             if (response.isCompletedExceptionally()) {
                 negotiation.transitionConfirming();
                 negotiationStore.save(negotiation);
+                invokeForEach(l -> l.confirming(negotiation));
                 monitor.debug(format("[Provider] Failed to send contract agreement with id %s. ContractNegotiation %s stays in state %s.",
                         agreement.getId(), negotiation.getId(), ContractNegotiationStates.from(negotiation.getState())));
                 continue;
```

Comment:
uhm... there are a lot of `continue`, maybe this refactor deserves another issue to avoid overwhelming this?

Reply:
What's so bad/dangerous about `continue` statements? But sure, could be refactored in another PR.

Reply:
they lead to a not-so-readable code flow, at least for an hi-level programming language. In this case, for example, an `else` condition would give better meaning on what the code it's doing.

Looking better at this code, the real error is using the `isCompletedExceptionally` on a future that could be still not completed, better use the `whenComplete`, otherwise that condition could never be meet. Will open an issue for this, sorry for the flooding

Reply:
Makes sense, thanks for opening the issue.